File: leetcode/p00033/p00033.go
package p00033

// Time Complexity: O(logn)
// Space Complexity: O(1)
func search(nums []int, target int) int {
	head, tail := 0, len(nums)-1
	for head <= tail {
		mid := (head + tail) / 2
		if nums[mid] == target {
			return mid
		}

		if nums[mid] < nums[tail] {
			// 如果右邊有序, 且 target 也在這個範圍內
			if nums[mid] < target && nums[tail] >= target {
				head = mid + 1
			} else {
				// 不然就丟到另外一邊
				tail = mid - 1
			}
		} else {
			// 左邊有序, 且 target 也在這個範圍
			if nums[head] <= target && nums[mid] > target {
				tail = mid - 1
			} else {
				head = mid + 1
			}
		}
	}
	return -1
}

File: leetcode/p00033/p00033.py
from typing import List

class Solution:
    # Time Complexity: O(logn)
    # Space Complexity: O(1)
    def search(self, nums: List[int], target: int) -> int:
        left = 0
        right = len(nums)-1

        while left <= right:
            mid = left + (right - left) // 2
            if nums[mid] == target:
                return mid
            
            if nums[mid] >= nums[left]:
                if target < nums[left] or target > nums[mid]:
                    left = mid + 1
                else:
                    right = mid - 1
            else:
                if target < nums[mid] or target > nums[right]:
                    right = mid - 1
                else:
                    left = mid + 1

        return -1

if __name__ == '__main__':
    s = Solution()

    assert s.search([4,5,6,7,0,1,2], 0) == 4
    assert s.search([4,5,6,7,0,1,2], 3) == -1
    assert s.search([1], 0) == -1
    assert s.search([3,1], 1) == 1